Spherical (3D-looking) `Markers` symbols

brisvag opened the pull request:
Turns out getting #2204 into the `Markers` visual was doable. This is an alternative implementation of that PR: ~instead of creating a new visual, I used the infrastructure of `Markers` and added another symbol called `sphere`. In pratice, this symbol uses the same cropping function as `disc`, but adds all the lighting stuff that #2204 introduced.~

This implements a `spherical` attribute in the `MarkersVisual` that add a sphere-like shading to the markers. It's applicable to all the symbols, but particularly useful with `disc`, because it results in depth-correct spheres.

Due to the significant changes to the shaders, this PR would in practice also supersede #2198.

This is the render from the added example:

https://user-images.githubusercontent.com/23482191/133228329-f1bc3482-dac5-40a7-ac3d-8c6ba036b211.mp4

## Differences
- ~Compared to the other implementation, this shoudl make it easier for people to convert markers into spheres. However, some minor changes to the api might be introduced. Nothing is exactly breaking, but since marker sizes were inconsistent between scaling and non-scaling mode, some examples might break, especially due to the performance issue when zoomed in a lot.~
- This PR also has antialiasing (albeit ugly when markers overlap, see #2208).
- Due to using markers, this PR also allows to show an `edge`.

## Questions
- I'm not a huge fan of how `set_data` reinitializes everything unless specified. This means that to update one parameter we need to pass all the previous values manually. Should I split out some things from it (e.g: anything "global"?). My feeling is that the default should be that nothing gets changed (all `None`?)
- antialias is kind of ok in 2D, if not many points overlap, but is basically unusable in 3D. Should we default to off until #2039 is merged?

djhoese commented on examples/scene/marker_spheres.py:
```diff
@@ -0,0 +1,49 @@
+"""
+Show some spheres and sticks
```
```suggestion
Spheres and Sticks
```

djhoese commented on examples/scene/point_cloud.py:
```diff
@@ -23,20 +23,21 @@
 
 # generate data
-pos = np.random.normal(size=(100000, 3), scale=0.2)
+pos = np.random.normal(size=(10000, 3), scale=0.2)
```
Are the changes in this example needed or was this accidentally committed?

brisvag replied:
Oops, reacted to the wrong thing by mistake. This particular change was accidental (no performance hit that big, no worries! :)). There are a couple of changes that were on purpose, specifically on the scaling, due to some inconsistencies. I still need to debug exactly what is going on.

brisvag replied:
Ok, everything is fixed now, no changes needed.

djhoese commented on vispy/visuals/markers.py:
```diff
@@ -27,95 +28,169 @@
+    // gl_PointSize is the diameter
+    gl_PointSize = $v_size + 4. * (v_edgewidth + 1.5 * u_antialias);
+
+    if (u_sphere == true) {
```
Could this be named and treated in a more generic way? Like maybe this isn't just `sphere` but is more of a "3D markers" effect? So `u_3d` or something? Overall you are finding some Z distance right?

brisvag replied:
I did think about this, but then focused on getting this to work first :P I think it's actually doable and not too much work... whtether it will look good, it's another question :P The result would be that any marker would look like a "carved out" sphere that always faces the screen. Anything beyond a sphere radius would look terrible, though, because it's flattened out to the same depth and lighting value. Maybe that can be solved as well, but it's definitely harder.

brisvag replied:
I did it! It was trivial, in the end :) It does look pretty decent on most markers (square ones are a bit weird on the edges, but oh well).


https://user-images.githubusercontent.com/23482191/131989560-dbf11cc6-1528-4c91-80f9-778dbac5c84b.mp4

djhoese reviewed (commented):
I had a couple comments inline, but otherwise this looks very interesting. Thanks for trying to get this to work. I'm in a privileged position here where I don't understand most of what is required at a low-level for this to work but I know enough to make manager-style requests that completely ruin your day. So here's my newest one:

It seems that most of this new functionality has to do with lighting. Could this be separated out to an optional (maybe embedded) lighting Filter class? Even if it can only be used for Markers maybe that is OK. Just throwing it out there.

brisvag commented:
In general, I think what you're asking for *can* be done, but I'mnot sure it's worth the effort to solve all the issues that come from non-spherical spheres :P. I'll play around a bit, and come back with some isnights!

brisvag commented:
There's a thing I find odd about the `edge` (which already existed but is now more obvious due to the spheres symbol):

https://user-images.githubusercontent.com/23482191/131884461-0948ade1-3fef-4608-aa22-dc2284f76513.mp4

the edge is actually half on and half off the marker. This is obviously intended, but I wonder why; I would either put it completely outside or completely inside the marker. While this hardly matters when `scaling=False`, it's misleading when `scaling=True`, because the size of the point is no longer right in data coordinates (see #2198 for a case where this is relevant).

In my opinion, this should be changed to either be render *outside* the markers, to preserve the absolute size of the *face* when changing the edge width, or inside so the sum of face+edge is preserved. The second option would also be better for the sphere case, where we can define a reasonable z value for the edge.

djhoese commented:
I know this isn't helpful, but I'm not sure I understand the new shader code enough to have an opinion on this. We can always wait for @almarklein, @asnt, @kmuehlbauer, @larsoner, and @rougier to comment if they have any opinions. I think if the overall behavior of the existing markers is relatively preserved but the behavior of the edge is objectively-ish improved then that's fine by me. It seems like everything you are identifying is being improved by what you propose and is likely not something that the original MarkersVisual was concerned about or was introduced when we started worrying about Markers interacting with other visuals (like the depth test change).

rougier commented:
The halo stuff you see most probably comes from the antialiased gradient we use to delineate the marker. From memory half othe linewidth is inside the marker and the other half is outside such that the antialiased gradient is more of less coincidental with the marker shape. If you use markers for sphere, you need probably to cancel the outer border (linewidth=0) or else you'll aliasing artifacts as it is the case with markers.

brisvag commented:
Not sure which halo you're referring to... If you mean what's shown more obviously in #2208, then yes, that's the antialias. Here, however, the white border you see is just the edge width, aliasing is off. Due to the fact that, as you said, the *edge* is half on and half off the marker (just like the aliasing), the marker size is not preserved. This is quite significant when the edge is more than, say, a pixel (see example above where the edge is half the size as the marker).

This is no big deal if you don't care abotu the absolute size of the marker, but it's a problem if you do (marker size is effectively `size + 0.5 * edge_width`). Additionally, this causes the ugly consequences above when using this PR's sphere symbol, because the frag depth very obviously highlights this size inconsistency.

I image initially the marker edge was conceived as a "1-sh pixel border`, but I've seen it used many times differently. Having said that, yes, one can simply remove the edge and ignore this if needed :)

brisvag commented on vispy/visuals/markers.py:
```diff
@@ -27,95 +28,170 @@
     // [-e/2+a, e/2-a] core edge (center 0, diameter e-2a = 2t)
     // [e/2-a, e/2+a] antialising edge-background
-    float t = 0.5*v_edgewidth - v_antialias;
+    // use max because we don't want negative transition zone
+    float t = max(0.5*v_edgewidth - u_antialias, 0);
```
Actually, removing the edge completely is impossible in master, but I changed it here.

rougier commented:
Sorry, did not see your other post. For the half/half stroke part, think this is more or less a standard way to proceed and might ba part of the SVG specification (but I'm not sure).

brisvag commented:
Updated so any symbol can be made `spherical`. I used this rather than `3D` for clarity, and because attribute names can't start with numbers :P

I made some significant changes to the `add_data` function, removing everything that's non marker-dependent and moving it into a property (and the init).

This means that `symbol` also was moved out, since you can't set symbols on a per-marker basis. I think it makes more sense like this, but it might be too much of an API change to go without deprecation; what does anyone think?

rougier commented:
If you have some coding time left, you could also add "cylindrical". I think this is a standard way to render molecules (see for example http://www.sunsetlakesoftware.com/2011/05/08/enhancing-molecules-using-opengl-es-20)

brisvag commented:
Wow, this is cool, thanks for the link!! It looks like it's quit e abit more work; maybe I'll get to it at some point, but not for now :)

rougier commented:
For the record: http://vcg.isti.cnr.it/Publications/2006/TCM06/Tarini_FinalVersionElec.pdf

brisvag commented:
I'm not sure how to deal with the errors in the tests. I think what's happening is that the deprecation errors I introduced are causing the failure... Should I change `GraphVisual` so that it uses the properties for `symbol` and `scaling`  instead of `set_data`?

sofroniewn commented:
I love this!! Will be awesome to get this added.

As to one day having cylinders ❤️ !! That will be so nice. Then I think we could definitely start adding some molecular visualization to napari 🚀

djhoese commented:
> Should I change GraphVisual so that it uses the properties for symbol and scaling instead of set_data?

@brisvag I'm not sure. Even though you have the deprecation warnings you still actually set them. The error makes it seem like some shader property/attribute isn't being set, right? Updating the GraphVisual to do things correctly would be good either way though.

brisvag commented:
Not sure what's going on... The test run is just stuck there for a long time, and then is marked as "cancelled", but counts as failed.

djhoese commented:
It still has the same error:

```
----------------------------------------------------------------------
Example /basics/visuals/graph.py failed (0):
----------------------------------------------------------------------
WARNING: Error drawing visual <vispy.visuals.markers.MarkersVisual object at 0x7f3b9c80ebd0>
WARNING: Traceback (most recent call last):
  File "<string>", line 17, in <module>
  File "/home/runner/work/vispy/vispy/examples/basics/visuals/graph.py", line 36, in __init__
    self.show()
  File "/home/runner/work/vispy/vispy/vispy/app/canvas.py", line 438, in show
    self._backend._vispy_set_visible(visible)
  File "/home/runner/work/vispy/vispy/vispy/app/backends/_qt.py", line 471, in _vispy_set_visible
    self.showNormal()
  File "/home/runner/work/vispy/vispy/vispy/app/backends/_qt.py", line 567, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/runner/work/vispy/vispy/vispy/app/backends/_qt.py", line 567, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/home/runner/work/vispy/vispy/vispy/app/backends/_qt.py", line 897, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/home/runner/work/vispy/vispy/vispy/util/event.py", line 453, in __call__
    self._invoke_callback(cb, event)
  File "/home/runner/work/vispy/vispy/vispy/util/event.py", line 473, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/home/runner/work/vispy/vispy/vispy/util/event.py", line 469, in _invoke_callback
    cb(event)
  File "/home/runner/work/vispy/vispy/examples/basics/visuals/graph.py", line 50, in on_draw
    self.visual.draw()
  File "/home/runner/work/vispy/vispy/vispy/visuals/visual.py", line 605, in draw
    v.draw()
  File "/home/runner/work/vispy/vispy/vispy/visuals/visual.py", line 452, in draw
    self._vshare.index_buffer)
  File "/home/runner/work/vispy/vispy/vispy/visuals/shaders/program.py", line 102, in draw
    Program.draw(self, *args, **kwargs)
  File "/home/runner/work/vispy/vispy/vispy/gloo/program.py", line 495, in draw
    raise RuntimeError('Must have at least one attribute')
RuntimeError: Must have at least one attribute
ERROR: Invoking <bound method Canvas.on_draw of <Canvas (PyQt5) at 0x7f3ba91d2090>> for DrawEvent
WARNING: Error drawing visual <vispy.visuals.markers.MarkersVisual object at 0x7f3b9c80ebd0>
```

djhoese commented:
Are you able to run the example locally?

brisvag commented:
No; I thought I did but I somehow was on a different branch, silly me :P Will check up on this.

brisvag commented:
I'm very confused... `examples/scene/graph.py` works fine, and `examples/basics/visuals/graph.py` fails with that error, despite the 2 having basically the same setup in terms of parameters. I'm not familiar with the scene subclassing approach, so maybe I'm missing something obvious in the examples that fails. Can someone take a look?

djhoese commented:
@brisvag So debugging it one step at a time I found two things:

1. The exception is checking `_user_variables` that are of the type `DataBuffer`. If you print (or put a breakpoint) at the exception you'll see that for the graph example that doesn't work there are only numpy arrays. If you do it for the one that works (the line just above the exception) then you'll see that the position (vertex buffer) is set. If you run the bad example with breakpoints at all the points in the GraphVisual where `self._nodes.set_data` is called you'll notice they aren't called. I'm not sure if this a problem with the example code or your changes, but...
2. You seem to use `_node_properties` in the graph for both the set of properties to look for and the dictionary of property values to hold onto. But even when I tried to fix it I still didn't get `set_data` called on the MarkersVisual inside GraphVisual.

Let me know if you get any further.

brisvag commented:
Ok, I got it. Due to the changes I made, no vertex buffer was being set when instantiating `Markers` unless `data` was passed to it. This is all fine, except that on draw, this failed. To fix it, I added a check (see 5c95e71) identical to that for `markers._symbol` so that draw is skipped unless data is present. Does that seem reasonable?

djhoese commented:
That seems like such an obvious thing to include and I'm a little concerned that it wasn't earlier. Works for me.

brisvag commented:
CI still hangs, but with no error this time O.o

djhoese commented:
Restarting in the hope that it was an accident. Assuming this failure is an example script hanging then it may still be possible to fix it.

brisvag commented:
The example `magnify.py` is always getting stuck, but works locally... no error, again :/

djhoese commented:
Hm even running `python -m make test examples examples/scene/magnify.py` works fine on my local system. Going to be difficult to figure this out if we can't debug locally. @brisvag what do you think about commenting out the `Markers` usage in `magnify.py` and see if that lets the tests pass?

djhoese commented:
Found it! It isn't magnify.py it is `examples/scene/marker_spheres.py` that hangs. Debugging now...

djhoese commented:
So what is the current state of this? @almarklein are you OK with this? @brisvag how does this make the existing markers look by default? Does it change their appearance at all?

brisvag commented:
There are a couple of changes to default markers:
1) fixed what @kevinyamauchi was working on in #2198, which was incorporated directly here. This means that now perspective and zooming work as intended both with scaling enabled and disabled.
2) `edge_width` are now self-consistend depending on mode scaling/non-scaling modes (consequence of 1). This means that if scaling is enabled, `edge_width` is in *data pixels* (just like `size`), and not in *screen pixels*. This seems better to me, but maybe others think otherwise. I like that if I set `size` to `10` and `edge_width` to `1`, the result is a 10 to 1 ratio no matter the zoom/scaling settings.
3) `DeprecationWarning`s are raised when passing `symbol` or `scaling` to `set_data`; now, they should be set with the respective properties.
4) added properties for previously hidden settings like `antialias` and `alpha`, plus all the new things (`light_color`, 'spherical`...)

djhoese commented:
Ok I think that all sounds good to me.

almarklein commented on vispy/visuals/markers.py:
```diff
@@ -611,21 +736,87 @@ def symbol(self, symbol):
 
+    @property
+    def scaling(self):
+        return self._scaling
```
docstring :)

almarklein replied:
Also the other props below.

almarklein commented on vispy/visuals/markers.py:
```diff
@@ -611,21 +736,87 @@ def symbol(self, symbol):
+
+    @scaling.setter
+    def scaling(self, value):
+        self.shared_program['u_scaling'] = value
```
Maye cast `value` to bool here for good measure.

almarklein commented on vispy/visuals/markers.py:
```diff
@@ -611,21 +736,87 @@ def symbol(self, symbol):
+
+    @antialias.setter
+    def antialias(self, value):
+        self.shared_program['u_antialias'] = value
```
Maybe cast `value` to float here.

almarklein reviewed (changes requested):
A few comments, mostly about cleaning up these new props.

> This means that if scaling is enabled, edge_width is in data pixels (just like size), and not in screen pixels. 

It feels to me that `edge_width` and `size` should indeed be in the same space.

djhoese commented:
@almarklein has a better understanding of this PR than I do so his approval makes me think I approve it too. Plus if anything goes badly I can just blame him :stuck_out_tongue_winking_eye:

djhoese commented:
Thanks @brisvag!
